File: Golang/15.三数之和.go
package leetcode

import "sort"

func threeSum(nums []int) [][]int {
	var ans [][]int
	var st map[[3]int]bool = make(map[[3]int]bool)
	sort.Ints(nums)
	for i := 0; i < len(nums)-2; i++ {
		var target int = -nums[i]
		left, right := i+1, len(nums)-1
		for left < right {
			if nums[left]+nums[right] == target {
				if _, ok := st[[3]int{-target, nums[left], nums[right]}]; !ok {
					st[[3]int{-target, nums[left], nums[right]}] = true
					ans = append(ans, []int{-target, nums[left], nums[right]})
				}
				left++
				right--
			} else if nums[left]+nums[right] < target {
				left++
			} else {
				right--
			}
		}
	}
	return ans
}

File: TypeScript/15.三数之和.ts
function threeSum(nums: number[]): number[][] {
    if (nums.length < 3)
        return []
    nums.sort((x, y) => x - y)
    const ans: number[][] = new Array<number[]>()
    const st: Set<string> = new Set<string>()
    for (let i = 0; i < nums.length - 2; ++i) {
        const target = -nums[i]
        let [left, right] = [i + 1, nums.length - 1]
        while (left < right) {
            if (nums[left] + nums[right] == target) {
                if (!st.has(JSON.stringify([-target, nums[left], nums[right]]))) {
                    st.add(JSON.stringify([-target, nums[left], nums[right]]))
                    ans.push([-target, nums[left], nums[right]])
                }
                ++left
                --right
            } else if (nums[left] + nums[right] < target) {
                ++left
            } else {
                --right
            }
        }
    }
    return ans
};
